hw/ip/otbn/dv/tracer/rtl/otbn_tracer.sv

```
@@ -0,0 +1,229 @@
+ * it provides to the simulation environment via a DPI call, it uses `otbn_trace_intf` to obtain the
+ * information it needs.
+ *
+ * The simulation environment must provide a `accept_otbn_trace_string` (see defintion in RTL below)
+ * function exported via DPI. Every cycle there is trace output the tracer will call this function
+ * (at most once per clock cycle) providing the trace output with a cycle count.
+ *
+ * Trace output consists of one or more lines. Each line begins with a single character that
+ * idenitfies the category of output followed by a space. The format following that depends upon the
+ * category (see the XPrefix parameters below for the possible categories).
+ *
+ * Ordering amongst the lines in a single cycle isn't guaranteed though it is guaranteed that if the
+ * tracer will output an 'E' or 'S' (instruction execute or stall) line in a cycle that will be the
+ * first line of the trace for that cycle.
+ */
+module otbn_tracer
+(
+  input  logic  clk_i,
+  input  logic  rst_ni,
+
+  otbn_trace_intf otbn_trace
+);
+  import otbn_pkg::*;
+
+  parameter string RegReadPrefix = "<";
+  parameter string RegWritePrefix = ">";
+  parameter string InsnExecutePrefix = "E";
+  parameter string InsnStallPrefix = "S";
+  parameter string MemWritePrefix = "W";
+  parameter string MemReadPrefix = "R";
+
+  string trace_output_buffer;
+
+  logic [31:0] cycle_count;
+
+  // Given a WLEN size word output a hex string with the data split into 32-bit chunks seperated
+  // with '_'
+  function automatic string otbn_wlen_data_str(logic [WLEN-1:0] data);
+    string data_str;
+    bit first;
+
+    first = 1;
+
+    for (int i = WLEN; i > 0; i-= 32) begin
+      if (!first) begin
+        data_str = $sformatf("%s_", data_str);
+      end else begin
+        data_str = "0x";
+        first = 0;
+      end
+
+      if (i >= 32) begin
+        data_str = $sformatf("%s%08x", data_str, data[i-1 -: 32]);
+      end
```

Is this check to cope with WLEN not being a multiple of 32? Can't we use an assertion instead? At the moment, I think you silently drop data in that case.

---

That was the plan, I did indeed forget the else clause. Sadly verilator doesn't like my fixed version

```
if (i >= 32) begin
  data_str = $sformatf("%s%08x", data_str, data[i-1 -: 32]);
end else 
  data_str = $sformatf("%s%08x", data_str, data[i-1 : 0]);
end
```

I've just dropped the check entirely for now. I should determine what verilator does when you don't have a multiple of 32 (and indeed what the system verilog spec says should happen).

I could add an assertion though was uncertain of the state of verilator support for assert statements.

---

Verilator is fine with immediate asserts. I guess it complains about `data[i-1:0]` because it doesn't know the size in advance. I think you could probably do something like define a "tail_length" outside of the loop using something like `WLEN % 32` and then use `data[i-1 -: tail_length]` (I haven't tried this).

---

I've gone for an immediate assert, looking at the LRM slicing does indeed need to be of constant width (so `[i - 1 : 0]` is probably disallowed, of course the LRM doesn't go into details about 'constant', one could argue that given the parameters of the for loop the i is indeed sufficiently constant for the purposes of statically determining the width of the slice).

---

Well, that's why I suggested something based on `WLEN % 32`, which you could assign to a `localparam` (which is a constant, as defined by the spec; it's a bit tricky to nail down, but I think the relevant bit is section 6.20 plus the slightly bizarre `constant_foo` entries in the BNF, which try to define something obviously non-syntactical, ho hum!)

---

I felt that non multiple of 32 bit WLEN sizes probably aren't something we're interested in so better to just keep things simple. Easy enough to improve the function to support them if we do find such support is needed and the assert should quickly clue us into the fact this is needed if we ever attempt it.